src/value.c

```diff
@@ -412,12 +412,34 @@ size_t SIValue_StringConcat(SIValue* strings, unsigned int string_count, char *b
 }
 
 int SIValue_Compare(SIValue a, SIValue b) {
-  // TODO: Use SI_NUMERIC.
-  if(a.type == T_DOUBLE) {
-    return a.doubleval - b.doubleval;
-  } else {
-    return strcasecmp(a.stringval, b.stringval);
+  // Types are identical
+  if (a.type == b.type) {
+    if (a.type == T_DOUBLE) {
+      /* TODO - inf, -inf, and  NaN do not behave canonically in this routine,
+       * though they do not break anything. Low-priority bug. */
+      return a.doubleval - b.doubleval;
+    } else if (a.type == T_STRING) {
+      return strcasecmp(a.stringval, b.stringval);
+    } else {
+      // TODO matching unhandled types - revisit this when introducing other numerics
+      return 0;
+    }
+  }
+
+  // Types differ
+  /* Cypher specifies the following ascending order of disjoint types:
+   * - String
+   * - Number
+   * - NULL
+   */
+  if (a.type == T_STRING) {
+    return -1;
   }
+  if (b.type == T_STRING) {
+    return 1;
+  }
+
+  return a.type - b.type;
```

Comment:
Delta between types?

Reply:
Yeah, it looks a little silly, but it allows us to always favor a defined type over a T_NULL type. Actually, the value of T_STRING is 0x1 and T_DOUBLE is 0x40 - if we modified that so that T_STRING was larger, we could omit the above if/else too.

Reply:
Is there a documentation by Neo4J which explains the above?

Reply:
Need to document this within SIValue types.

Reply:
Yes, I used the Cypher Query Language Reference v9, the *Orderability* section. Specifically:

```
we propose a pre-determined order of types and ensure that each value falls
under exactly one disjoint type in this order. We define the following ascending global sort order of disjoint types:
• MAP types
• Regular map
• NODE
• RELATIONSHIP
• LIST OF ANY?
• PATH
• STRING
• BOOLEAN
• NUMBER
• NaN values are treated as the largest numbers in orderability only (i.e. they are put after
positive infinity)
• VOID (i.e. the type of null)
```